clients/hadoopfs/src/main/java/io/lakefs/auth/AWSLakeFSTokenProvider.java

```diff
@@ -108,15 +107,39 @@ public Request<GeneratePresignGetCallerIdentityRequest> newPresignedRequest() th
 
     public String newPresignedGetCallerIdentityToken() throws Exception {
         Request<GeneratePresignGetCallerIdentityRequest> signedRequest = this.newPresignedRequest();
+        Map<String, ?> rawQueryParams = signedRequest.getParameters();
+        Map<String, String> params = new HashMap<>();
+        // check if the value is an array and join it with commas depends on the AWS SDK version
+        for (Map.Entry<String, ?> entry : rawQueryParams.entrySet()) {
+            String key = entry.getKey();
+            Object value = entry.getValue();
+            if (value instanceof String[]) {
+                String[] arrayValue = (String[]) value;
+                String joinedValue = String.join(",", arrayValue);
+                params.put(key, joinedValue);
+            } else {
+                params.put(key, value.toString());
+            }
+        }
+
         // generate token parameters object
-        JSONObject identityTokenParams = new JSONObject();
-        identityTokenParams.put("method", signedRequest.getHttpMethod().name());
-        identityTokenParams.put("endpoint", signedRequest.getEndpoint().toString());
-        identityTokenParams.put("signedHeaders", signedRequest.getHeaders().keySet().toArray());
-        identityTokenParams.put("expiration", this.stsExpirationInSeconds);
-        identityTokenParams.put("signedParams", signedRequest.getParameters());
+        LakeFSExternalPrincipalIdentityRequest identityTokenParams = new LakeFSExternalPrincipalIdentityRequest(
+                signedRequest.getHttpMethod().name(),
+                signedRequest.getEndpoint().getHost(),
+                this.stsPresigner.getSignedRegion(signedRequest),
+                params.get(STSGetCallerIdentityPresigner.AMZ_ACTION_PARAM_NAME),
+                params.get(STSGetCallerIdentityPresigner.AMZ_DATE_PARAM_NAME),
+                params.get(STSGetCallerIdentityPresigner.AMZ_EXPIRES_PARAM_NAME),
+                this.awsProvider.getCredentials().getAWSAccessKeyId(),
+                params.get(STSGetCallerIdentityPresigner.AMZ_SIGNATURE_PARAM_NAME),
+                Arrays.asList(params.get(STSGetCallerIdentityPresigner.AMZ_SIGNED_HEADERS_PARAM_NAME).split(";")),
+                params.get(STSGetCallerIdentityPresigner.AMZ_VERSION_PARAM_NAME),
+                params.get(STSGetCallerIdentityPresigner.AMZ_ALGORITHM_PARAM_NAME),
+                params.get(STSGetCallerIdentityPresigner.AMZ_SECURITY_TOKEN_PARAM_NAME)
+        );
+
         // base64 encode
-        return Base64.encodeBase64String(identityTokenParams.toString().getBytes());
+        return Base64.encodeBase64String(identityTokenParams.toJSON().getBytes());
```

Comment:
I don't understand.  AFAIK JSON encoding is not unique.  Don't we need to force particular ordering of parameters, particular whitespacing, etc.?

Reply:
Not at all, it contains all the parameters that we need to assemble the URL. 
Only one of them requires order and it's already a list. 
The signed headers, it will preserve it.